### PR TITLE
Kyle Krueger - Add Address Retrieve Logic

### PR DESCRIPTION
**Pull request by kkrueger-bravo**
Testing Evidence - API Call
image

Testing Evidence - Logs
image

### CODE REVIEW

File: src/test/java/com/bravo/user/service/AddressServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+		this.dtoAddresses = ids.stream().map(id -> createAddressDto(Integer.toString(id)))
+				.collect(Collectors.toList());
+
+		when(resourceMapper.convertAddresses(daoAddresses)).thenReturn(dtoAddresses);
+	}
+
+	@Test
+	void retrieveByUserId() {
+		String userId = "123a-456b";
```
Comment (wheelers-websites) [inline src/test/java/com/bravo/user/service/AddressServiceTest.java:60]: `userId` could be `final` here.

File: src/test/java/com/bravo/user/controller/AddressControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+		String userId = "123a-456b";
+
+		when(addressService.retrieveByUserId(anyString())).thenReturn(addresses);
+
+		final ResultActions result = this.mockMvc.perform(get("/address/retrieve/".concat(userId)))
+				.andExpect(status().isOk());
+
+		for (int i = 0; i < addresses.size(); i++) {
+			result.andExpect(
+					jsonPath(String.format("$[%d].id", i)).value(addresses.get(i).getId()));
```
Comment (wheelers-websites) [inline src/test/java/com/bravo/user/controller/AddressControllerTest.java:63]: Weird formatting here.

File: src/main/java/com/bravo/user/controller/AddressController.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	public AddressController(AddressService addressService, UserValidator userValidator) {
+		this.addressService = addressService;
+		this.userValidator = userValidator;
+	}
+
+	@GetMapping(value = "/retrieve/{userId}")
+	@ResponseBody
+	public List<AddressDto> retrieve(final @PathVariable String userId,
+			final @RequestParam(required = false) Integer page,
+			final @RequestParam(required = false) Integer size,
```
Comment (wheelers-websites) [inline src/main/java/com/bravo/user/controller/AddressController.java:34]: You're accepting the `page` and `size` parameters, but not using them.

File: src/main/java/com/bravo/user/service/AddressService.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+	private final AddressRepository addressRepository;
+	private final ResourceMapper resourceMapper;
+
+	public AddressService(AddressRepository addressRepository, ResourceMapper resourceMapper) {
+		this.addressRepository = addressRepository;
+		this.resourceMapper = resourceMapper;
+	}
+
+	public List<AddressDto> retrieveByUserId(final String userId) {
+		List<Address> addressList = addressRepository.findByUserId(userId);
```
Comment (wheelers-websites) [inline src/main/java/com/bravo/user/service/AddressService.java:28]: `addressList` can be `final` here.

File: src/test/java/com/bravo/user/controller/AddressControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+	public void beforeEach() {
+		final List<Integer> ids = IntStream.range(1, 10).boxed().collect(Collectors.toList());
+
+		this.addresses = ids.stream().map(id -> createAddressDto(Integer.toString(id)))
+				.collect(Collectors.toList());
+	}
+
+	@Test
+	void getRetrieveByUserId() throws Exception {
+		String userId = "123a-456b";
```
Comment (wheelers-websites) [inline src/test/java/com/bravo/user/controller/AddressControllerTest.java:54]: `userId` can be `final` here.

File: src/main/java/com/bravo/user/service/AddressService.java (hunk 4 above)
Comment (tlynema-bravolt) [reply src/main/java/com/bravo/user/service/AddressService.java:28]: Does final on a list make it immutable -- for Kyle?

**Comment by wheelers-websites** [comment]
I want to start by saying that we are not expecting any additional commits. Feel free to make them if you wish, but that's not an expectation that we have. With that said I will go through your changes and leave technical comments on them.

File: src/test/java/com/bravo/user/dao/model/mapper/ResourceMapperTest.java [hunk 6]
```diff
@@ -1,30 +1,26 @@
...
+import com.bravo.user.dao.model.Address;
+import com.bravo.user.dao.model.User;
+import com.bravo.user.model.dto.AddressDto;
+import com.bravo.user.model.dto.UserReadDto;
 
 @ContextConfiguration(classes = {App.class})
 @ExtendWith(SpringExtension.class)
 @SpringBootTest
-public class ResourceMapperTest {
+class ResourceMapperTest {
```
Comment (tlynema-bravolt) [inline src/test/java/com/bravo/user/dao/model/mapper/ResourceMapperTest.java:23]: Bonus points for this :)

File: src/test/java/com/bravo/user/service/AddressServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+	private ResourceMapper resourceMapper;
+
+	@MockBean
+	private AddressRepository addressRepository;
+
+	private List<AddressDto> dtoAddresses;
+
+	@BeforeEach
+	public void beforeEach() {
+		final List<Integer> ids = IntStream.range(1, 10).boxed().collect(Collectors.toList());
```
Comment (tlynema-bravolt) [inline src/test/java/com/bravo/user/service/AddressServiceTest.java:45]: I like streams.  This falls as a non-judicious use IMHO https://dev.to/kylec32/effective-java-use-stream-judiciously-4178.

File: src/main/java/com/bravo/user/controller/AddressController.java (hunk 3 above)
Comment (kkrueger-bravo) [reply src/main/java/com/bravo/user/controller/AddressController.java:34]: good catch, I initially was going to implement pagination until I re-read the requirements to simply return all addresses for the userId. I will clean them up, as well as the unused httpResponse parameter

File: src/main/java/com/bravo/user/service/AddressService.java (hunk 4 above)
Comment (kkrueger-bravo) [reply src/main/java/com/bravo/user/service/AddressService.java:28]: @wheelerswebservices - in my past experience, I haven't been so liberal with specifying variables final, but after some research it sounds like this is beneficial in a few ways. IDEs are even able to implement missing keywords automatically as a Save action- thanks for the advice

@tlynema-bravolt - no, making a List (or any object) final does not make it's contents immutable, it simply locks in the reference

File: src/test/java/com/bravo/user/controller/AddressControllerTest.java (hunk 5 above)
Comment (kkrueger-bravo) [reply src/test/java/com/bravo/user/controller/AddressControllerTest.java:54]: thanks- will fix

File: src/test/java/com/bravo/user/controller/AddressControllerTest.java (hunk 2 above)
Comment (kkrueger-bravo) [reply src/test/java/com/bravo/user/controller/AddressControllerTest.java:63]: this was initially auto-formatted via Eclipse, but I will manually adjust this to make it single-lined to improve readability

File: src/test/java/com/bravo/user/service/AddressServiceTest.java (hunk 7 above)
Comment (kkrueger-bravo) [reply src/test/java/com/bravo/user/service/AddressServiceTest.java:45]: My goal was to remain as similar to the development and test approach to the User components (which is using this already). In order to adjust, I'm curious what are your thoughts though on why this is non-judicious? Is it not explicit enough or should it just be handled in a more straightforward way?

Would you recommend I hardcode a list of ids to test with instead?

File: src/test/java/com/bravo/user/service/AddressServiceTest.java (hunk 1 above)
Comment (kkrueger-bravo) [reply src/test/java/com/bravo/user/service/AddressServiceTest.java:60]: thanks- will fix

File: src/test/java/com/bravo/user/service/AddressServiceTest.java (hunk 7 above)
Comment (tlynema-bravolt) [reply src/test/java/com/bravo/user/service/AddressServiceTest.java:45]: > Would you recommend I hardcode a list of ids to test with instead?

Was thinking about this too.  A simple for loop would work here with a list.add.  The end result is to get the same List of integers and this code does that before every test.  This could be run once and used for all tests that way. 

I googled it to learn this one myself :) https://stackoverflow.com/questions/10242380/how-can-i-generate-a-list-or-array-of-sequential-integers-in-java  has variations with performance implications.  In this case, I think both of the lines could be combined into IntStream.rangeClosed(1, 10).forEach(x -> logic)  where the logic in the next line would be  within the stream.  That has the benefit of the range not being stored in a collection which is a bit better on memory.  

The code in the codebase is sometimes not perfect :) .   One of the goals is to see things that you pick up that we miss.  You did this in the test code which was awesome!

TLDR instead of stream -> List -> stream  go with just stream for the logic in lines 46-49.
